src/main/java/org/candlepin/resource/CloudRegistrationResource.java

```diff
@@ -83,4 +156,149 @@ private CloudRegistrationData getCloudRegistrationData(CloudRegistrationDTO clou
+            CloudAuthTokenType.STANDARD :
+            CloudAuthTokenType.ANONYMOUS;
+
+        return new CloudAuthenticationResultDTO()
+            .ownerKey(ownerKey)
+            .anonymousConsumerUuid(anonymousConsumerUuid)
+            .token(token)
+            .tokenType(tokenType.toString());
+    }
+
+    /**
+     * Validates the values of in {@link CloudAuthenticationResult}.
+     *
+     * @param result
+     *  the {@link CloudAuthenticationResult} to validate
+     *
+     * @throws CloudRegistrationAuthorizationException
+     *  if a required field does no meet the validation requirements
+     */
+    private void validateCloudAuthenticationResult(CloudAuthenticationResult result) {
+        String cloudAccountId = result.getCloudAccountId();
+        if (cloudAccountId == null || cloudAccountId.isBlank()) {
+            String errmsg = this.i18n.tr("cloud account ID could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String cloudInstanceId = result.getCloudInstanceId();
+        if (cloudInstanceId == null || cloudInstanceId.isBlank()) {
+            String errmsg = this.i18n.tr("cloud instance ID could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        CloudProvider cloudProvider = result.getCloudProvider();
+        if (cloudProvider == null) {
+            String errmsg = this.i18n.tr("cloud provider could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String offerId = result.getOfferId();
+        if (offerId == null || offerId.isBlank()) {
+            String errmsg = this.i18n.tr("offer ID could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String productId = result.getProductId();
+        if (productId == null || productId.isBlank()) {
+            String errmsg = this.i18n.tr("product ID could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+    }
+
+    /**
+     * Determines if an owner exists and that owner has a pool for the product ID.
+     *
+     * @param ownerKey
+     *  the key to the owner that is being verified
+     *
+     * @param productId
+     *  the ID of the product that is being verified
+     *
+     * @return
+     *  true if the owner exists and has a pool for the product ID, or false if the owner
+     *  does not exist or does not have a pool for the product ID.
+     */
+    private boolean doesOwnerAndPoolExist(String ownerKey, String productId) {
+        boolean ownerAndPoolExists = false;
+        if (ownerKey != null && !ownerKey.isBlank()) {
+            Owner owner = ownerCurator.getByKey(ownerKey);
+            if (owner != null) {
+                ownerAndPoolExists = !poolCurator.listByOwnerAndProduct(owner, productId).isEmpty();
```

Comment:
We had talked about this before as a double query optimization, but looking at the actual underlying query here, we should probably write something custom specifically for this lookup rather than doing two individual queries and all the builder pain that goes into that `listByOwnerAndProduct` call.

Specifically, I believe the product ID in this case is going to be a SKU-level product, so we can really optimize the query down to a couple of joins. Given that this will probably be getting a lot of usage, it'll be better than hitting one of the slowest queries in CP repeatedly. Something on the order of:

```
// rename me pls
PoolCurator.poolWithProductExistsInOrg(owner_key, product_id):

  String jpql = "SELECT pool.id FROM Pool pool JOIN pool.owner owner JOIN pool.product prod WHERE owner.key = :owner_key AND prod.id = :product_id";

  // remembering to set the result limit to 1 to short circuit out
  return !this.getEntityManager()
    .createQuery(jpql)
    .setParameter("owner_key", owner_key)
    .setParameter("product_id", product_id)
    .setMaxResults(1)
    .getResultList()
    .isEmpty();
```

Reply:
Added this new method `hasPoolForProduct` in the PoolCurator for this!